### ISSUE TITLE
curl: (60) SSL certificate problem: unable to get local issuer certificate

### ISSUE THREAD

**Issue by vasiliygagin** [git-for-windows/git#4084]
- [ ] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?

```
$ git --version --build-options

git version 2.38.1.windows.1
cpu: x86_64
built from commit: b85c8f604d375d4d773a36842964e8a7ec056aae
sizeof-long: 4
sizeof-size_t: 8
shell-path: /bin/sh
feature: fsmonitor--daemon
```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

```
$ cmd.exe /c ver

Microsoft Windows [Version 10.0.22000.1098]
```

 - What options did you set as part of the installation? Or did you choose the
   defaults? defaults

```
# One of the following:
> type "C:\Program Files\Git\etc\install-options.txt"
> type "C:\Program Files (x86)\Git\etc\install-options.txt"
> type "%USERPROFILE%\AppData\Local\Programs\Git\etc\install-options.txt"
> type "$env:USERPROFILE\AppData\Local\Programs\Git\etc\install-options.txt"
$ cat /etc/install-options.txt

Editor Option: VIM
Custom Editor Path:
Default Branch Option:
Path Option: Cmd
SSH Option: OpenSSH
Tortoise Option: false
CURL Option: OpenSSL
CRLF Option: CRLFAlways
Bash Terminal Option: MinTTY
Git Pull Behavior Option: Merge
Use Credential Manager: Enabled
Performance Tweaks FSCache: Enabled
Enable Symlinks: Disabled
Enable Pseudo Console Support: Disabled
Enable FSMonitor: Disabled
```

 - Any other interesting things about your environment that might be related
   to the issue you're seeing?

No

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

Git Bash

 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

```
curl -v https://raw.githubusercontent.com/nvm-sh/nvm/v0.39.2/install.sh
```
 - What did you expect to occur after running these commands?

file being downloaded

 - What actually happened instead?

```
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0*   Trying 185.199.110.133:443...
* Connected to raw.githubusercontent.com (185.199.110.133) port 443 (#0)
* ALPN: offers h2
* ALPN: offers http/1.1
*  CAfile: C:/Program Files/Git/mingw64/ssl/certs/ca-bundle.crt
*  CApath: none
} [5 bytes data]
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
} [512 bytes data]
* TLSv1.3 (IN), TLS handshake, Server hello (2):
{ [122 bytes data]
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
{ [15 bytes data]
* TLSv1.3 (IN), TLS handshake, Certificate (11):
{ [1151 bytes data]
* TLSv1.3 (OUT), TLS alert, unknown CA (560):
} [2 bytes data]
* SSL certificate problem: unable to get local issuer certificate
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
* Closing connection 0
curl: (60) SSL certificate problem: unable to get local issuer certificate
More details here: https://curl.se/docs/sslcerts.html

curl failed to verify the legitimacy of the server and therefore could not
establish a secure connection to it. To learn more about this situation and
how to fix it, please visit the web page mentioned above.
```

It seems that cert file used "ca-bundle.crt" has proper certificate for "DigiCert Global Root CA"
curl should not be failing.

**Comment by dscho** [comment]
I cannot reproduce this issue:

```console
$ curl -v -o /dev/null https://raw.githubusercontent.com/nvm-sh/nvm/v0.39.2/install.sh
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0*   Trying 2606:50c0:8003::154:443...
* Connected to raw.githubusercontent.com (2606:50c0:8003::154) port 443 (#0)
* ALPN: offers h2
* ALPN: offers http/1.1
*  CAfile: C:/Program Files/Git/mingw64/ssl/certs/ca-bundle.crt
*  CApath: none
} [5 bytes data]
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
} [512 bytes data]
* TLSv1.3 (IN), TLS handshake, Server hello (2):
{ [122 bytes data]
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
{ [19 bytes data]
* TLSv1.3 (IN), TLS handshake, Certificate (11):
{ [3052 bytes data]
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
{ [264 bytes data]
* TLSv1.3 (IN), TLS handshake, Finished (20):
{ [52 bytes data]
* TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
} [1 bytes data]
* TLSv1.3 (OUT), TLS handshake, Finished (20):
} [52 bytes data]
* SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384
* ALPN: server accepted h2
* Server certificate:
*  subject: C=US; ST=California; L=San Francisco; O=GitHub, Inc.; CN=*.github.com
*  start date: Apr  7 00:00:00 2022 GMT
*  expire date: Apr  7 23:59:59 2023 GMT
*  subjectAltName: host "raw.githubusercontent.com" matched cert's "*.githubusercontent.com"
*  issuer: C=US; O=DigiCert Inc; CN=DigiCert TLS RSA SHA256 2020 CA1
*  SSL certificate verify ok.
* Using HTTP2, server supports multiplexing
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
} [5 bytes data]
* h2h3 [:method: GET]
* h2h3 [:path: /nvm-sh/nvm/v0.39.2/install.sh]
* h2h3 [:scheme: https]
* h2h3 [:authority: raw.githubusercontent.com]
* h2h3 [user-agent: curl/7.85.0]
* h2h3 [accept: */*]
* Using Stream ID: 1 (easy handle 0x16dfc6e0330)
} [5 bytes data]
> GET /nvm-sh/nvm/v0.39.2/install.sh HTTP/2
> Host: raw.githubusercontent.com
> user-agent: curl/7.85.0
> accept: */*
>
{ [5 bytes data]
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
{ [193 bytes data]
< HTTP/2 200
< cache-control: max-age=300
< content-security-policy: default-src 'none'; style-src 'unsafe-inline'; sandbox
< content-type: text/plain; charset=utf-8
< etag: "6eac97d42f3e93feb7f0ba3e444b543f31a98f4fc0bb2231347a4607545dabc0"
< strict-transport-security: max-age=31536000
< x-content-type-options: nosniff
< x-frame-options: deny
< x-xss-protection: 1; mode=block
< x-github-request-id: 0803:5719:64634A:702D8B:635B131D
< accept-ranges: bytes
< date: Fri, 28 Oct 2022 07:49:59 GMT
< via: 1.1 varnish
< x-served-by: cache-hhn4078-HHN
< x-cache: HIT
< x-cache-hits: 1
< x-timer: S1666943400.746731,VS0,VE1
< vary: Authorization,Accept-Encoding,Origin
< access-control-allow-origin: *
< x-fastly-request-id: 75e6559070cc5afba86d46ec7b1e6ba0da645bd5
< expires: Fri, 28 Oct 2022 07:54:59 GMT
< source-age: 288
< content-length: 15916
<
{ [5 bytes data]
100 15916  100 15916    0     0   374k      0 --:--:-- --:--:-- --:--:--  388k
* Connection #0 to host raw.githubusercontent.com left intact
```

**Comment by vasiliygagin** [comment]
Thank you for checking.
Just in case, I have ran MD5 sums. Can you please compare to yours?

> md5sum.exe ca-bundle.crt
> 194e94a736822099695d5b96a80f1044 *ca-bundle.crt
> 
> $ md5sum.exe /mingw64/bin/curl
> 6f5fbf3f8c30bcad03f275bb58216486 */mingw64/bin/curl
>

**Comment by dscho** [comment]
@vasiliygagin checks out all right:

```console
$ md5sum.exe /mingw64/ssl/certs/ca-bundle.crt /mingw64/bin/curl
194e94a736822099695d5b96a80f1044 */mingw64/ssl/certs/ca-bundle.crt
6f5fbf3f8c30bcad03f275bb58216486 */mingw64/bin/curl
```

**Comment by dscho** [comment]
Are you maybe running behind some kind of a proxy?

**Comment by dscho** [comment]
There are many ways a proxy could be configured. I'm not sure whether it would show up in `curl -v`'s output.

**Comment by dscho** [comment]
No, I did not mean the upper-case `V` (as in "Version") but the lower-case `v` (as in "verbose"). In any case, you will have to look around for yourself how to detect the proxy. It could be a `https_proxy` environment variable. Or something in Internet Options. Or some config option. Or something else entirely.

**Comment by dscho** [comment]
No, I did not mean the upper-case `V` (as in "Version") but the lower-case `v` (as in "verbose"). In any case, you will have to look around for yourself how to detect the proxy. It could be a `https_proxy` environment variable. Or something in Internet Options. Or some config option. Or something else entirely.

**Comment by dscho** [comment]
> that was [here.](https://github.com/git-for-windows/git/issues/4084#issuecomment-1312178687) It was command `curl -v https://raw.githubusercontent.com/nvm-sh/nvm/v0.39.2/install.sh`

Well, it said in its output: "Funkce zrušení nemohla zkontrolovat zrušení certifikátu.". Google Translate claims that this is Czech, but I do not speak nor understand that language. It's kind of hard to help if it is made deliberately hard.

**Comment by vasiliygagin** [comment]
No proxy here

**Comment by dscho** [comment]
I've only ever encountered 0x80092012 in the context of a MITM proxy (such as Fiddler), see e.g. https://stackoverflow.com/questions/54938026/curl-unknown-error-0x80092012-the-revocation-function-was-unable-to-check-r.

In Git for Windows, we do have 720ec872af7aa4624883e041ea2d463e65d06582 to address these issues, and to make that work, we contributed https://github.com/curl/curl/pull/4981 which was accepted.

If you use `curl` directly, you will want to use the [`--ssl-revoke-best-effort` option](https://curl.se/docs/manpage.html#--ssl-revoke-best-effort).

In any case, we strayed _really_ far away from any possible _Git_ bug, therefore I will close this ticket.
